ext/standard/array.c

```diff
@@ -3825,6 +3825,59 @@ PHPAPI int php_array_replace_recursive(HashTable *dest, HashTable *src) /* {{{ *
 }
 /* }}} */
 
+/* Returns true if it's possible to do an in-place array modification, preventing a costly copy.
+ * It also modifies the CV (if applicable) to prevent freeing it upon assigning.
+ * If this returns true you need to add a ref at the end of the modification for the return value.
+ * safe_for_CV indicates whether the optimization may be applied to CVs. This should be false
+ * if the operation may throw for example, because then the in-place modification is visible to
+ * the exception catcher. */
+static bool prepare_in_place_array_modify_if_possible(const zend_execute_data *execute_data, const zval *arg, bool safe_for_CV)
+{
+	/* 2 refs: the CV and the argument; or 1 ref for a temporary passed as argument */
+	uint32_t refcount = Z_REFCOUNT_P(arg);
+	if (refcount != 1 && !(refcount == 2 && safe_for_CV)) {
+		return false;
+	}
+	/* Immutable or persistent => no modification allowed */
+	if (GC_FLAGS(Z_ARRVAL_P(arg)) & (IS_ARRAY_IMMUTABLE | IS_ARRAY_PERSISTENT)) {
+		return false;
+	}
+
+	if (refcount == 2) {
+		const zend_op *call_opline = execute_data->prev_execute_data->opline;
+		const zend_op *next_opline = call_opline + 1;
+
+		/* Must be an assignment from the result of the call to a CV */
+		if (next_opline->opcode != ZEND_ASSIGN || next_opline->op1_type != IS_CV || next_opline->op2.var != call_opline->result.var) {
+			return false;
+		}
+
+		/* Must be an assignment to the same array as the input */
+		zval *var = ZEND_CALL_VAR(execute_data->prev_execute_data, next_opline->op1.var);
+		if (Z_TYPE_P(var) != IS_ARRAY || Z_ARRVAL_P(arg) != Z_ARRVAL_P(var)) {
+			return false;
+		}
+
+		/* Must set the CV to NULL so we don't destroy the array on assignment */
+		ZVAL_NULL(var);
```

Comment:
Sorry, I posted wrong example above, and modified it in-place.

Reply:
> Can var be typed at this place? (typed reference)

I don't think this is a concern. `array_merge` and `array_merge_recursive` accept the parameter by-value. This means that we'll get a refcount of 2 for the array (while keeping the reference at 1), at which point this optimization will not apply.

> In case of exception the value of CV is going to have an improper value.

This is a valid concern, we thought about coercion warnings but missed destructor side-effects. I can't think of a way to avoid this.